controlpanel/api/aws.py

```diff
@@ -24,248 +22,415 @@ def iam_arn(resource, account=settings.AWS_ACCOUNT_ID):
     return arn("iam", resource, account=account)
 
 
-class AWSClient(object):
-    def __init__(self):
-        self.enabled = settings.ENABLED["write_to_cluster"]
-        self.client = boto3.client
+BASE_ROLE_POLICY = {
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "Service": "ec2.amazonaws.com",
+            },
+            "Action": "sts:AssumeRole",
+        },
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "AWS": iam_arn(f"role/{settings.K8S_WORKER_ROLE_NAME}"),
+            },
+            "Action": "sts:AssumeRole",
+        },
+    ],
+}
+
+OIDC_STATEMENT = {
+    "Effect": "Allow",
+    "Principal": {
+        "Federated": iam_arn(f"oidc-provider/{settings.OIDC_DOMAIN}"),
+    },
+    "Action": "sts:AssumeRoleWithWebIdentity",
+}
+
+SAML_STATEMENT = {
+    "Effect": "Allow",
+    "Principal": {
+        "Federated": iam_arn(f"saml-provider/{settings.SAML_PROVIDER}"),
+    },
+    "Action": "sts:AssumeRoleWithSAML",
+    "Condition": {
+        "StringEquals": {
+            "SAML:aud": "https://signin.aws.amazon.com/saml",
+        },
+    },
+}
+
+READ_INLINE_POLICIES = f'{settings.ENV}-read-user-roles-inline-policies'
+
+BASE_S3_ACCESS_POLICY = {
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Sid": "console",
+            "Action": [
+                "s3:GetBucketLocation",
+                "s3:ListAllMyBuckets",
+            ],
+            "Effect": "Allow",
+            "Resource": ["arn:aws:s3:::*"],
+        },
+    ],
+}
+
+BASE_S3_ACCESS_STATEMENT = {
+    'list': {
+        'Sid': 'list',
+        'Action': [
+            's3:ListBucket',
+        ],
+        'Effect': 'Allow',
+    },
+    'readonly': {
+        'Sid': 'readonly',
+        'Action': [
+            's3:GetObject',
+            's3:GetObjectAcl',
+            's3:GetObjectVersion',
+        ],
+        'Effect': 'Allow',
+    },
+}
+
+BASE_S3_ACCESS_STATEMENT['readwrite'] = {
```

Comment:
I've only started to have a look (so it may be a stupid observation) but I found the fact that we're changing a "constant" a bit confusing.
Also the fact that the `BASE_S3_ACCESS_STATEMENT` already contains the `readonly` and `readwrite` statements could lead to bugs if we're not careful.

Although I get that without resources it wouldn't actually grant any permission.

But let me have a look at the rest to see how this is used.

Reply:
This dict is just used as a lookup for constant values - perhaps it would be better as `BASE_LIST_STATEMENT`, `BASE_READONLY_STATEMENT` and `BASE_READWRITE_STATEMENT`. Having them in a dict lets me look up the base statement for the SID programmatically later on.

Reply:
Changing to your suggestion - it's much better.